File: 01-ll-designs/message_queue/solutions/cpp/message_queue.cpp
/**
 * Simple In-Memory Message Queue Implementation in C++
 * 
 * This implementation provides a thread-safe message queue system with support for:
 * - Multiple topics with concurrent access
 * - Producer-consumer patterns with multiple subscribers
 * - FIFO message ordering within topics
 * - Subscription management with dynamic subscribe/unsubscribe
 * - Statistics and monitoring capabilities
 */

#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <memory>
#include <mutex>
#include <thread>
#include <chrono>
#include <atomic>
#include <functional>
#include <algorithm>
#include <future>
#include <random>
#include <sstream>

/**
 * Generates a simple UUID-like string
 */
std::string generateId() {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<> dis(0, 15);
    
    std::stringstream ss;
    for (int i = 0; i < 8; ++i) {
        ss << std::hex << dis(gen);
    }
    return ss.str();
}

/**
 * Represents a message in the queue
 */
struct Message {
    std::string id;
    std::string topic;
    std::string payload;
    std::chrono::system_clock::time_point timestamp;
    std::unordered_map<std::string, std::string> headers;
    
    Message(const std::string& topic, const std::string& payload, 
            const std::unordered_map<std::string, std::string>& headers = {})
        : id(generateId()), topic(topic), payload(payload), 
          timestamp(std::chrono::system_clock::now()), headers(headers) {}
    
    std::string toString() const {
        return "Message{id='" + id.substr(0, 8) + "', topic='" + topic + 
               "', payload='" + payload + "'}";
    }
};

/**
 * Abstract interface for handling received messages
 */
class MessageHandler {
public:
    virtual ~MessageHandler() = default;
    virtual void handleMessage(const Message& message) = 0;
};

// Forward declarations
class MessageQueue;
class Topic;

/**
 * Message consumer that can subscribe to topics
 */
class Consumer {
private:
    std::string id_;
    std::unique_ptr<MessageHandler> handler_;
    std::unordered_set<std::string> subscribedTopics_;
    std::atomic<bool> active_;
    mutable std::mutex topicsMutex_;

public:
    Consumer(const std::string& id, std::unique_ptr<MessageHandler> handler)
        : id_(id), handler_(std::move(handler)), active_(true) {}
    
    ~Consumer() {
        stop();
    }
    
    void onMessage(const Message& message) {
        if (active_.load()) {
            try {
                handler_->handleMessage(message);
            } catch (const std::exception& e) {
                std::cerr << "Error processing message " << message.id << ": " << e.what() << std::endl;
            }
        }
    }
    
    void stop() {
        active_.store(false);
    }
    
    bool isActive() const {
        return active_.load();
    }
    
    const std::string& getId() const { return id_; }
    
    std::unordered_set<std::string> getSubscribedTopics() const {
        std::lock_guard<std::mutex> lock(topicsMutex_);
        return subscribedTopics_;
    }
    
    void addSubscription(const std::string& topic) {
        std::lock_guard<std::mutex> lock(topicsMutex_);
        subscribedTopics_.insert(topic);
    }
    
    void removeSubscription(const std::string& topic) {
        std::lock_guard<std::mutex> lock(topicsMutex_);
        subscribedTopics_.erase(topic);
    }
};

/**
 * Statistics structure for topics
 */
struct TopicStats {
    std::string name;
    size_t messageCount;
    size_t queueSize;
    size_t subscriberCount;
    size_t maxSize;
};

/**
 * Represents a topic with its subscribers and message queue
 */
class Topic {
private:
    std::string name_;
    size_t maxSize_;
    std::queue<Message> messages_;
    std::vector<std::shared_ptr<Consumer>> subscribers_;
    std::atomic<size_t> messageCount_;
    mutable std::mutex messagesMutex_;
    mutable std::mutex subscribersMutex_;

public:
    Topic(const std::string& name, size_t maxSize = 1000)
        : name_(name), maxSize_(maxSize), messageCount_(0) {}
    
    void addMessage(const Message& message) {
        {
            std::lock_guard<std::mutex> lock(messagesMutex_);
            if (messages_.size() >= maxSize_) {
                std::cerr << "Topic " << name_ << " is full, dropping message: " << message.id << std::endl;
                return;
            }
            messages_.push(message);
            messageCount_.fetch_add(1);
        }
        deliverMessage(message);
    }
    
    void subscribe(std::shared_ptr<Consumer> consumer) {
        std::lock_guard<std::mutex> lock(subscribersMutex_);
        auto it = std::find(subscribers_.begin(), subscribers_.end(), consumer);
        if (it == subscribers_.end()) {
            subscribers_.push_back(consumer);
            consumer->addSubscription(name_);
        }
    }
    
    void unsubscribe(std::shared_ptr<Consumer> consumer) {
        std::lock_guard<std::mutex> lock(subscribersMutex_);
        auto it = std::find(subscribers_.begin(), subscribers_.end(), consumer);
        if (it != subscribers_.end()) {
            subscribers_.erase(it);
            consumer->removeSubscription(name_);
        }
    }
    
    TopicStats getStats() const {
        std::lock_guard<std::mutex> msgLock(messagesMutex_);
        std::lock_guard<std::mutex> subLock(subscribersMutex_);
        
        return TopicStats{
            name_,
            messageCount_.load(),
            messages_.size(),
            subscribers_.size(),
            maxSize_
        };
    }
    
    const std::string& getName() const { return name_; }

private:
    void deliverMessage(const Message& message) {
        std::vector<std::shared_ptr<Consumer>> currentSubscribers;
        {
            std::lock_guard<std::mutex> lock(subscribersMutex_);
            currentSubscribers = subscribers_;
        }
        
        // Deliver messages asynchronously to avoid blocking
        for (auto& subscriber : currentSubscribers) {
            if (subscriber && subscriber->isActive()) {
                // Use async to deliver message without blocking
                std::async(std::launch::async, [subscriber, message]() {
                    try {
                        subscriber->onMessage(message);
                    } catch (const std::exception& e) {
                        std::cerr << "Error delivering message to " << subscriber->getId() 
                                  << ": " << e.what() << std::endl;
                    }
                });
            } else if (subscriber) {
                // Remove inactive subscribers
                unsubscribe(subscriber);
            }
        }
    }
};

/**
 * Main message queue broker
 */
class MessageQueue {
private:
    std::unordered_map<std::string, std::shared_ptr<Topic>> topics_;
    std::vector<std::shared_ptr<Consumer>> consumers_;
    mutable std::mutex topicsMutex_;
    mutable std::mutex consumersMutex_;

public:
    std::shared_ptr<Topic> createTopic(const std::string& name, size_t maxSize = 1000) {
        std::lock_guard<std::mutex> lock(topicsMutex_);
        auto it = topics_.find(name);
        if (it == topics_.end()) {
            auto topic = std::make_shared<Topic>(name, maxSize);
            topics_[name] = topic;
            return topic;
        }
        return it->second;
    }
    
    bool deleteTopic(const std::string& name) {
        std::lock_guard<std::mutex> topicsLock(topicsMutex_);
        auto it = topics_.find(name);
        if (it != topics_.end()) {
            auto topic = it->second;
            topics_.erase(it);
            
            // Unsubscribe all consumers
            std::lock_guard<std::mutex> consumersLock(consumersMutex_);
            for (auto& consumer : consumers_) {
                if (consumer) {
                    topic->unsubscribe(consumer);
                }
            }
            return true;
        }
        return false;
    }
    
    std::string publish(const std::string& topicName, const std::string& payload,
                       const std::unordered_map<std::string, std::string>& headers = {}) {
        auto topic = createTopic(topicName);
        Message message(topicName, payload, headers);
        std::string messageId = message.id;
        topic->addMessage(message);
        return messageId;
    }
    
    void subscribe(std::shared_ptr<Consumer> consumer, const std::string& topicName) {
        auto topic = createTopic(topicName);
        topic->subscribe(consumer);
        
        std::lock_guard<std::mutex> lock(consumersMutex_);
        auto it = std::find(consumers_.begin(), consumers_.end(), consumer);
        if (it == consumers_.end()) {
            consumers_.push_back(consumer);
        }
    }
    
    void unsubscribe(std::shared_ptr<Consumer> consumer, const std::string& topicName) {
        std::lock_guard<std::mutex> lock(topicsMutex_);
        auto it = topics_.find(topicName);
        if (it != topics_.end()) {
            it->second->unsubscribe(consumer);
        }
    }
    
    std::optional<TopicStats> getTopicStats(const std::string& topicName) const {
        std::lock_guard<std::mutex> lock(topicsMutex_);
        auto it = topics_.find(topicName);
        if (it != topics_.end()) {
            return it->second->getStats();
        }
        return std::nullopt;
    }
    
    std::unordered_map<std::string, TopicStats> getAllTopicStats() const {
        std::lock_guard<std::mutex> topicsLock(topicsMutex_);
        std::lock_guard<std::mutex> consumersLock(consumersMutex_);
        
        std::unordered_map<std::string, TopicStats> stats;
        for (const auto& pair : topics_) {
            stats[pair.first] = pair.second->getStats();
        }
        return stats;
    }
    
    size_t getTopicCount() const {
        std::lock_guard<std::mutex> lock(topicsMutex_);
        return topics_.size();
    }
    
    size_t getConsumerCount() const {
        std::lock_guard<std::mutex> lock(consumersMutex_);
        return consumers_.size();
    }
};

/**
 * Message producer
 */
class Producer {
private:
    std::string id_;
    std::shared_ptr<MessageQueue> messageQueue_;

public:
    Producer(const std::string& id, std::shared_ptr<MessageQueue> messageQueue)
        : id_(id), messageQueue_(messageQueue) {}
    
    std::string publish(const std::string& topic, const std::string& payload,
                       const std::unordered_map<std::string, std::string>& headers = {}) {
        return messageQueue_->publish(topic, payload, headers);
    }
    
    const std::string& getId() const { return id_; }
};

/**
 * Simple message handler implementation for demonstration
 */
class PrintMessageHandler : public MessageHandler {
private:
    std::string consumerId_;

public:
    PrintMessageHandler(const std::string& consumerId) : consumerId_(consumerId) {}
    
    void handleMessage(const Message& message) override {
        std::cout << "[" << consumerId_ << "] Received message " 
                  << message.id.substr(0, 8) << " on topic '" << message.topic 
                  << "': " << message.payload << std::endl;
    }
};

/**
 * Demonstration function
 */
void demo() {
    std::cout << "=== Message Queue Demo ===" << std::endl << std::endl;
    
    // Create message queue
    auto mq = std::make_shared<MessageQueue>();
    
    // Create consumers
    auto consumer1 = std::make_shared<Consumer>("consumer-1", 
        std::make_unique<PrintMessageHandler>("consumer-1"));
    auto consumer2 = std::make_shared<Consumer>("consumer-2", 
        std::make_unique<PrintMessageHandler>("consumer-2"));
    auto consumer3 = std::make_shared<Consumer>("consumer-3", 
        std::make_unique<PrintMessageHandler>("consumer-3"));
    
    // Create producer
    Producer producer("producer-1", mq);
    
    // Subscribe consumers to topics
    std::cout << "Setting up subscriptions..." << std::endl;
    mq->subscribe(consumer1, "orders");
    mq->subscribe(consumer2, "orders");
    mq->subscribe(consumer3, "notifications");
    
    // Publish some messages
    std::cout << std::endl << "Publishing messages..." << std::endl;
    producer.publish("orders", "Order #1001 created");
    producer.publish("orders", "Order #1002 created");
    producer.publish("notifications", "System maintenance scheduled");
    producer.publish("orders", "Order #1003 created");
    
    // Wait for message processing
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    
    // Show statistics
    std::cout << std::endl << "=== Statistics ===" << std::endl;
    auto stats = mq->getAllTopicStats();
    for (const auto& pair : stats) {
        const auto& topicStats = pair.second;
        std::cout << "Topic '" << topicStats.name << "': " 
                  << topicStats.messageCount << " messages, " 
                  << topicStats.subscriberCount << " subscribers" << std::endl;
    }
    
    // Unsubscribe a consumer
    std::cout << std::endl << "Unsubscribing consumer-1 from orders..." << std::endl;
    mq->unsubscribe(consumer1, "orders");
    
    // Publish more messages
    std::cout << "Publishing more messages..." << std::endl;
    producer.publish("orders", "Order #1004 created");
    
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    std::cout << std::endl << "Demo completed!" << std::endl;
}

int main() {
    demo();
    return 0;
}

File: 01-ll-designs/message_queue/solutions/java/MessageQueue.java
/**
 * Simple In-Memory Message Queue Implementation in Java
 * 
 * This implementation provides a thread-safe message queue system with support for:
 * - Multiple topics with concurrent access
 * - Producer-consumer patterns with multiple subscribers
 * - FIFO message ordering within topics
 * - Subscription management with dynamic subscribe/unsubscribe
 * - Statistics and monitoring capabilities
 */

import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicLong;
import java.time.Instant;

/**
 * Represents a message in the queue
 */
class Message {
    private final String id;
    private final String topic;
    private final String payload;
    private final long timestamp;
    private final Map<String, String> headers;
    
    public Message(String topic, String payload, Map<String, String> headers) {
        this.id = UUID.randomUUID().toString();
        this.topic = topic;
        this.payload = payload;
        this.timestamp = Instant.now().toEpochMilli();
        this.headers = headers != null ? new HashMap<>(headers) : new HashMap<>();
    }
    
    // Getters
    public String getId() { return id; }
    public String getTopic() { return topic; }
    public String getPayload() { return payload; }
    public long getTimestamp() { return timestamp; }
    public Map<String, String> getHeaders() { return new HashMap<>(headers); }
    
    @Override
    public String toString() {
        return String.format("Message{id='%s', topic='%s', payload='%s'}", 
                           id.substring(0, 8), topic, payload);
    }
}

/**
 * Interface for handling received messages
 */
interface MessageHandler {
    void handleMessage(Message message);
}

/**
 * Message consumer that can subscribe to topics
 */
class Consumer {
    private final String id;
    private final MessageHandler handler;
    private final Set<String> subscribedTopics;
    private volatile boolean active;
    
    public Consumer(String id, MessageHandler handler) {
        this.id = id;
        this.handler = handler;
        this.subscribedTopics = ConcurrentHashMap.newKeySet();
        this.active = true;
    }
    
    public void onMessage(Message message) {
        if (active) {
            try {
                handler.handleMessage(message);
            } catch (Exception e) {
                System.err.println("Error processing message " + message.getId() + ": " + e.getMessage());
            }
        }
    }
    
    public void stop() {
        this.active = false;
    }
    
    public boolean isActive() {
        return active;
    }
    
    public String getId() { return id; }
    public Set<String> getSubscribedTopics() { return new HashSet<>(subscribedTopics); }
    
    void addSubscription(String topic) {
        subscribedTopics.add(topic);
    }
    
    void removeSubscription(String topic) {
        subscribedTopics.remove(topic);
    }
}

/**
 * Represents a topic with its subscribers and message queue
 */
class Topic {
    private final String name;
    private final int maxSize;
    private final BlockingQueue<Message> messages;
    private final List<Consumer> subscribers;
    private final AtomicLong messageCount;
    private final Object subscriberLock = new Object();
    
    public Topic(String name, int maxSize) {
        this.name = name;
        this.maxSize = maxSize;
        this.messages = new ArrayBlockingQueue<>(maxSize);
        this.subscribers = new ArrayList<>();
        this.messageCount = new AtomicLong(0);
    }
    
    public void addMessage(Message message) {
        try {
            if (messages.offer(message)) {
                messageCount.incrementAndGet();
                deliverMessage(message);
            } else {
                System.err.println("Topic " + name + " is full, dropping message: " + message.getId());
            }
        } catch (Exception e) {
            System.err.println("Error adding message to topic " + name + ": " + e.getMessage());
        }
    }
    
    public void subscribe(Consumer consumer) {
        synchronized (subscriberLock) {
            if (!subscribers.contains(consumer)) {
                subscribers.add(consumer);
                consumer.addSubscription(name);
            }
        }
    }
    
    public void unsubscribe(Consumer consumer) {
        synchronized (subscriberLock) {
            if (subscribers.remove(consumer)) {
                consumer.removeSubscription(name);
            }
        }
    }
    
    private void deliverMessage(Message message) {
        List<Consumer> currentSubscribers;
        synchronized (subscriberLock) {
            currentSubscribers = new ArrayList<>(subscribers);
        }
        
        // Deliver messages asynchronously to avoid blocking
        for (Consumer subscriber : currentSubscribers) {
            if (subscriber.isActive()) {
                CompletableFuture.runAsync(() -> subscriber.onMessage(message))
                    .exceptionally(throwable -> {
                        System.err.println("Error delivering message to " + subscriber.getId() + ": " + throwable.getMessage());
                        return null;
                    });
            } else {
                // Remove inactive subscribers
                unsubscribe(subscriber);
            }
        }
    }
    
    public Map<String, Object> getStats() {
        Map<String, Object> stats = new HashMap<>();
        synchronized (subscriberLock) {
            stats.put("name", name);
            stats.put("messageCount", messageCount.get());
            stats.put("queueSize", messages.size());
            stats.put("subscriberCount", subscribers.size());
            stats.put("maxSize", maxSize);
        }
        return stats;
    }
    
    public String getName() { return name; }
}

/**
 * Main message queue broker
 */
public class MessageQueue {
    private final ConcurrentHashMap<String, Topic> topics;
    private final List<Producer> producers;
    private final List<Consumer> consumers;
    private final Object listsLock = new Object();
    
    public MessageQueue() {
        this.topics = new ConcurrentHashMap<>();
        this.producers = new ArrayList<>();
        this.consumers = new ArrayList<>();
    }
    
    public Topic createTopic(String name, int maxSize) {
        return topics.computeIfAbsent(name, k -> new Topic(name, maxSize));
    }
    
    public Topic createTopic(String name) {
        return createTopic(name, 1000);
    }
    
    public boolean deleteTopic(String name) {
        Topic topic = topics.remove(name);
        if (topic != null) {
            // Unsubscribe all consumers
            synchronized (listsLock) {
                for (Consumer consumer : consumers) {
                    topic.unsubscribe(consumer);
                }
            }
            return true;
        }
        return false;
    }
    
    public String publish(String topicName, String payload, Map<String, String> headers) {
        Topic topic = topics.computeIfAbsent(topicName, k -> new Topic(topicName, 1000));
        Message message = new Message(topicName, payload, headers);
        topic.addMessage(message);
        return message.getId();
    }
    
    public String publish(String topicName, String payload) {
        return publish(topicName, payload, null);
    }
    
    public void subscribe(Consumer consumer, String topicName) {
        Topic topic = topics.computeIfAbsent(topicName, k -> new Topic(topicName, 1000));
        topic.subscribe(consumer);
        
        synchronized (listsLock) {
            if (!consumers.contains(consumer)) {
                consumers.add(consumer);
            }
        }
    }
    
    public void unsubscribe(Consumer consumer, String topicName) {
        Topic topic = topics.get(topicName);
        if (topic != null) {
            topic.unsubscribe(consumer);
        }
    }
    
    public Map<String, Object> getTopicStats(String topicName) {
        Topic topic = topics.get(topicName);
        return topic != null ? topic.getStats() : null;
    }
    
    public Map<String, Object> getAllStats() {
        Map<String, Object> stats = new HashMap<>();
        Map<String, Object> topicStats = new HashMap<>();
        
        for (Topic topic : topics.values()) {
            topicStats.put(topic.getName(), topic.getStats());
        }
        
        synchronized (listsLock) {
            stats.put("topics", topicStats);
            stats.put("totalTopics", topics.size());
            stats.put("totalConsumers", consumers.size());
            stats.put("totalProducers", producers.size());
        }
        
        return stats;
    }
    
    void addProducer(Producer producer) {
        synchronized (listsLock) {
            producers.add(producer);
        }
    }
}

/**
 * Message producer
 */
class Producer {
    private final String id;
    private final MessageQueue messageQueue;
    
    public Producer(String id, MessageQueue messageQueue) {
        this.id = id;
        this.messageQueue = messageQueue;
        messageQueue.addProducer(this);
    }
    
    public String publish(String topic, String payload, Map<String, String> headers) {
        return messageQueue.publish(topic, payload, headers);
    }
    
    public String publish(String topic, String payload) {
        return messageQueue.publish(topic, payload);
    }
    
    public String getId() { return id; }
}

/**
 * Simple message handler implementation for demonstration
 */
class PrintMessageHandler implements MessageHandler {
    private final String consumerId;
    
    public PrintMessageHandler(String consumerId) {
        this.consumerId = consumerId;
    }
    
    @Override
    public void handleMessage(Message message) {
        System.out.printf("[%s] Received message %s on topic '%s': %s%n",
                         consumerId, message.getId().substring(0, 8), 
                         message.getTopic(), message.getPayload());
    }
}

/**
 * Demonstration class
 */
class MessageQueueDemo {
    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== Message Queue Demo ===\n");
        
        // Create message queue
        MessageQueue mq = new MessageQueue();
        
        // Create consumers
        Consumer consumer1 = new Consumer("consumer-1", new PrintMessageHandler("consumer-1"));
        Consumer consumer2 = new Consumer("consumer-2", new PrintMessageHandler("consumer-2"));
        Consumer consumer3 = new Consumer("consumer-3", new PrintMessageHandler("consumer-3"));
        
        // Create producer
        Producer producer = new Producer("producer-1", mq);
        
        // Subscribe consumers to topics
        System.out.println("Setting up subscriptions...");
        mq.subscribe(consumer1, "orders");
        mq.subscribe(consumer2, "orders");
        mq.subscribe(consumer3, "notifications");
        
        // Publish some messages
        System.out.println("\nPublishing messages...");
        producer.publish("orders", "Order #1001 created");
        producer.publish("orders", "Order #1002 created");
        producer.publish("notifications", "System maintenance scheduled");
        producer.publish("orders", "Order #1003 created");
        
        // Wait for message processing
        Thread.sleep(1000);
        
        // Show statistics
        System.out.println("\n=== Statistics ===");
        Map<String, Object> stats = mq.getAllStats();
        @SuppressWarnings("unchecked")
        Map<String, Object> topics = (Map<String, Object>) stats.get("topics");
        
        for (Map.Entry<String, Object> entry : topics.entrySet()) {
            @SuppressWarnings("unchecked")
            Map<String, Object> topicStats = (Map<String, Object>) entry.getValue();
            System.out.printf("Topic '%s': %d messages, %d subscribers%n",
                            entry.getKey(), 
                            topicStats.get("messageCount"),
                            topicStats.get("subscriberCount"));
        }
        
        // Unsubscribe a consumer
        System.out.println("\nUnsubscribing consumer-1 from orders...");
        mq.unsubscribe(consumer1, "orders");
        
        // Publish more messages
        System.out.println("Publishing more messages...");
        producer.publish("orders", "Order #1004 created");
        
        Thread.sleep(1000);
        System.out.println("\nDemo completed!");
    }
}